## Projet BigData/BI

Réalisé par : ABDENNROUI Ilyes / LAURENT Emmanuel / LEGRAND Antoine / RONGIER Thomas

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import scipy.stats as ss
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.pipeline import make_pipeline, make_union
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
import xgboost as xgb
pd.set_option('display.max_columns', None)

C:\Anaconda3\lib\site-packages\xgboost\compat.py:36: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import MultiIndex, Int64Index


## 1. Introduction 

Dans un contexte ou les talents travaillant dans le domaine de l'immobilier, sont de plus en plus rare, nous avons été mandaté par l'agence immobilière SIAD Immo pour trouver une solution à ce problème de recrutement à l'aide de technique de Machine Learning. Nous avons donc, en accord avec l'entreprise, décidé de mettre en place un outil permettant aux clients de faire estimer leurs biens de façon automatique moyennant une contrepartie financière. Cet outil permettra à l'entreprise de se passer d'une main d'oeuvre qui est de plus en plus difficile à recruter et permettra aux clients de gagner plus de temps. En effet, les estimations se feront de manière instantanée, il n'y aura plus de contrainte de rendez-vous ou de durée d'estimation.  

Pour nous permettre de réaliser notre projet, l'entreprise a mis à notre disposition un jeu de données contenant différentes informations sur 2919 logmements. Ces informations nous renseingent sur la surface du bien, l'état général, la date de construction... En tout et pour tout nous avons à disposition 81 variables.

Notre but, est de mettre en place un modèle de machine learning qui permet d'estimer le prix d'un bien en fonction des caractéristiques de ce dernier.

### Dictionnaire des données

In [2]:
dico = pd.read_excel('C:/Users/106443/Downloads/dictionnaire_de_données.xlsx')

In [5]:
dico

Nom champ Type de données  Longueur du champ   Contrainte  \
0              Id         Integer                NaN  Primary Key   
1      MSSubClass         Integer                NaN     Not Null   
2        MSZoning         Varchar               20.0     Not Null   
3     LotFrontage         Integer                NaN          NaN   
4         LotArea         Integer                NaN     Not Null   
..            ...             ...                ...          ...   
75         MoSold         Integer                NaN     Not Null   
76         YrSold         Integer                NaN     Not Null   
77       SaleType         Varchar               20.0     Not Null   
78  SaleCondition         Varchar               20.0     Not Null   
79      SalePrice         Integer                NaN     Not Null   

                                          Description  
0                             Identifiant de la ligne  
1                               La classe du bâtiment  
2                    La classe de la zone du bâtiment  
3   Distance en pied reliant la propriété à la rue...  
4              Taille de la propriété en pieds carrés  
..                                                ...  
75                                    Mois vendu (MM)  
76                              Année de vente (AAAA)  
77                                      Type de vente  
78                                 Condition de vente  
79                               Prix de l'habitation  

[80 rows x 5 columns]

### Importation des données

In [6]:
df_train = pd.read_csv("C:/Users/106443/Downloads/train.csv")
df_test = pd.read_csv("C:/Users/106443/Downloads/test.csv")
df_test_sub = pd.read_csv("C:/Users/106443/Downloads/sample_submission.csv")

In [7]:
#Dans le dataset de test, il manque la variable cible. Cette dernière, se trouve dans un autre dataset
#Il est donc nécessaire de faire une jointure entre ces 2 datasets
#Pour récupérer les informations
df_test = pd.merge(df_test, df_test_sub, how='inner', left_on = 'Id', right_on = 'Id')

In [8]:
#Sur Kaggle, le dataset a déjà été découpé en train et test set. 
#Néanmoins nous ne savons pas de quelle manière ce découpage a été fait.
#Nous allons donc concatener les 2 dataset pour réaliser le découpage nous même..
data = [df_train, df_test]
df = pd.concat(data)

In [9]:
#Apperçu de notre jeu de données après la concatenation 
df.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   
5         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
6         Lvl    AllPub    Inside       Gtl      Somerst       Norm   
7         Lvl    AllPub    Corner       Gtl       NWAmes       PosN   
8         Lvl    AllPub    Inside       Gtl      OldTown     Artery   
9         Lvl    AllPub    Corner       Gtl      BrkSide     Artery   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   
5       Norm     1Fam     1.5Fin            5            5       1993   
6       Norm     1Fam     1Story            8            5       2004   
7       Norm     1Fam     2Story            7            6       1973   
8       Norm     1Fam     1.5Fin            7            5       1931   
9     Artery   2fmCon     1.5Unf            5            6       1939   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   
5          1995     Gable  CompShg     VinylSd     VinylSd       None   
6          2005     Gable  CompShg     VinylSd     VinylSd      Stone   
7          1973     Gable  CompShg     HdBoard     HdBoard      Stone   
8          1950     Gable  CompShg     BrkFace     Wd Shng       None   
9          1950     Gable  CompShg     MetalSd     MetalSd       None   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   
5         0.0        TA        TA       Wood       Gd       TA           No   
6       186.0        Gd        TA      PConc       Ex       TA           Av   
7       240.0        TA        TA     CBlock       Gd       TA           Mn   
8         0.0        TA        TA     BrkTil       TA       TA           No   
9       

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2919 entries, 0 to 1458
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             2919 non-null   int64  
 1   MSSubClass     2919 non-null   int64  
 2   MSZoning       2915 non-null   object 
 3   LotFrontage    2433 non-null   float64
 4   LotArea        2919 non-null   int64  
 5   Street         2919 non-null   object 
 6   Alley          198 non-null    object 
 7   LotShape       2919 non-null   object 
 8   LandContour    2919 non-null   object 
 9   Utilities      2917 non-null   object 
 10  LotConfig      2919 non-null   object 
 11  LandSlope      2919 non-null   object 
 12  Neighborhood   2919 non-null   object 
 13  Condition1     2919 non-null   object 
 14  Condition2     2919 non-null   object 
 15  BldgType       2919 non-null   object 
 16  HouseStyle     2919 non-null   object 
 17  OverallQual    2919 non-null   int64  
 18  OverallC

In [13]:
df.shape

(2919, 81)

## 2. Description des données 

2919 observations et 81 variables caractérisant les biens immobiliers.

In [16]:
#La fonction describe nous permet de décrire de manière très rapide les différents informations 
#sur les variables quantitatives (moyenne, écart-type, val min et max...)
#Nous permettant ainsi de détecter d'enventuels problèmes.
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  2919.000000  2919.000000  2433.000000    2919.000000  2919.000000   
mean   1460.000000    57.137718    69.305795   10168.114080     6.089072   
std     842.787043    42.517628    23.344905    7886.996359     1.409947   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     730.500000    20.000000    59.000000    7478.000000     5.000000   
50%    1460.000000    50.000000    68.000000    9453.000000     6.000000   
75%    2189.500000    70.000000    80.000000   11570.000000     7.000000   
max    2919.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  2919.000000  2919.000000   2919.000000  2896.000000  2918.000000   
mean      5.564577  1971.312778   1984.264474   102.201312   441.423235   
std       1.113131    30.291442     20.894344   179.334253   455.610826   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1953.500000   1965.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1993.000000     0.000000   368.500000   
75%       6.000000  2001.000000   2004.000000   164.000000   733.000000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

        BsmtFinSF2    BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  \
count  2918.000000  2918.000000  2918.000000  2919.000000  2919.000000   
mean     49.582248   560.772104  1051.777587  1159.581706   336.483727   
std     169.205611   439.543659   440.766258   392.362079   428.701456   
min       0.000000     0.000000     0.000000   334.000000     0.000000   
25%       0.000000   220.000000   793.000000   876.000000     0.000000   
50%       0.000000   467.000000   989.500000  1082.000000     0.000000   
75%       0.000000   805.500000  1302.000000  1387.500000   704.000000   
max    1526.000000  2336.000000  6110.000000  5095.000000  2065.000000   

       LowQualFinSF    GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath  \
count   2919.000000  2919.000000   2917.000000   2917.000000  2919.000000   
mean       4.694416  1500.759849      0.429894      0.061364     1.568003   
std       46.396825   506.051045      0.524736      0.245687     0.552969   
min        0.000000   334.000000      0.000000      0.000000     0.000000   
25%        0.000000  1126.000000      0.000000      0.000000     1.000000   
50%        0.000000  1444.000000      0.000000      0.000000     2.000000   
75%        0.000000  1743.500000      1.000000      0.000000     2.000000   
max     1064.000000  5642.000000      3.000000      2.000000     4.000000   

          HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  \
count  2919.000000   2919.000000   2919.000000   2919.000000  2919.000000   
mean      0.380267      2.860226      1.044536      6.451524     0.597122   
std       0.502872      0.822693      0.214462      1.569379     0.646129   
min       0.000000      0.000000      0.000000      2.000000     0.000000   
25%       0.000000      2.000000      1.000000      5.000000     0.000000   
50%       0.000000      3.000000      1.000000      6.000000     1.000000   
75%       1.000000      3.000000      1.000000      7.000000     1.000000   
max       2.000000      8.000000      3.000000     15.000000     4.000000   

       GarageYrBlt   GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  \
count  2760.000000  2918.000000  2918.000000  2919.000000  2919.000000   
mean   1978.113406     1.766621   472.874572    93.709832    47.486811   
std      25.574285     0.761624   215.394815   126.526589    67.575493   
min    1895.000000     0.000000     0.000000     0.000000     0.000000   
25%    1960.000000     1.000000   320.000000     0.000000     0.000000   
50%    1979.000000     2.000000   480.000000     0.000000    26.000000   
75%    2002.000000     2.000000   576.000000   168.000000    70.000000   
max    2

## 3. Préparation des données

### Répartition des variables qualitativres

In [24]:
#Fréquence d'apparition des différentes modalités des variables qualitatives
#C'est à dire les variables de type 'object'

#Copie du dataset de base en ne gardant que les variables du type 'object'
df_object = df.select_dtypes(include=['object']).copy()
df_object.columns

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')

In [25]:
#Fréquence d'apparition de chaque modalité pour les variables qualitatives
for col in list(df_object.columns):
    print('la variable ' + col +' a '+ str(df_object[col].nunique()) +' modalités :\n ')
    print(df[col].value_counts(normalize=False).map(str) + ' soit: '+round(df[col].value_counts(normalize=True)*100,2).map(str)+'%')
    print('\n')

la variable MSZoning a 5 modalités :
 
RL         2265 soit: 77.7%
RM         460 soit: 15.78%
FV          139 soit: 4.77%
RH           26 soit: 0.89%
C (all)      25 soit: 0.86%
Name: MSZoning, dtype: object


la variable Street a 2 modalités :
 
Pave    2907 soit: 99.59%
Grvl       12 soit: 0.41%
Name: Street, dtype: object


la variable Alley a 2 modalités :
 
Grvl    120 soit: 60.61%
Pave     78 soit: 39.39%
Name: Alley, dtype: object


la variable LotShape a 4 modalités :
 
Reg    1859 soit: 63.69%
IR1     968 soit: 33.16%
IR2        76 soit: 2.6%
IR3       16 soit: 0.55%
Name: LotShape, dtype: object


la variable LandContour a 4 modalités :
 
Lvl    2622 soit: 89.83%
HLS      120 soit: 4.11%
Bnk      117 soit: 4.01%
Low       60 soit: 2.06%
Name: LandContour, dtype: object


la variable Utilities a 2 modalités :
 
AllPub    2916 soit: 99.97%
NoSeWa        1 soit: 0.03%
Name: Utilities, dtype: object


la variable LotConfig a 5 modalités :
 
Inside     2133 soit: 73.07%
Corner   

Les variables comme la variable "Utilities" n'ayant que 2 modalités dont l'une d'entre elles n'est présente qu'une seule occurence ne pourront pas être exploité dans notre modélisation 

### Renommer les variables : A FAIRE

In [ ]:
# On renomme les modalités ayant un nom peu explicite
df.rename(columns={'nombre_niveau': 'nbre_etage',
                   'tr002_type_batiment_description' : 'type_batiment',
                   'tr001_modele_dpe_type' : 'type_bien'
                   },
                   inplace=True)

### Les valeurs manquantes : A FAIRE

In [9]:
# Compter le nombre d'observations manquantes pour chaque colonne
missing_values = df.isna().sum()

# Afficher le nombre d'observations manquantes pour chaque colonne
print(missing_values)

missing_values_percent = round((df.isna().sum() / len(df)) * 100,2)
print(missing_values_percent.to_string())

Id                 0
MSSubClass         0
MSZoning           4
LotFrontage      486
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           1
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64
Id                0.00
MSSubClass        0.00
MSZoning          0.14
LotFrontage      16.65
LotArea           0.00
Street            0.00
Alley            93.22
LotShape          0.00
LandContour       0.00
Utilities         0.07
LotConfig         0.00
LandSlope         0.00
Neighborhood      0.00
Condition1        0.00
Condition2        0.00
BldgType          0.00
HouseStyle        0.00
OverallQual       0.00
OverallCond       0.00
YearBuilt         0.00
YearRemodAdd      0.00
RoofStyle         0.00
RoofMatl          0.00
Exterior1st       0.03
Exterior2nd       0.03
MasVnrType        0.82
MasVnrArea        0.79
ExterQual         0.00
ExterCond         0.00
Foundation        0.00
BsmtQual          2.77
BsmtCond          2.81
BsmtExpo

### Les valeurs aberrantes : A FAIRE

### Pour les variables quantitatives

In [41]:
#On récupère les variables quantitives 
df_int = df.select_dtypes(include=['int64','float64']).copy()
df_int.columns

Index(['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea',
       'MiscVal', 'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')

In [44]:
# Detection des valeurs aberrantes avec la méthode du Z score
for col in list(df_int.columns):
    mean = np.mean(df_int[col])
    std = np.std(df_int[col])

    threshold = 3
    outlier = []
    for i in df[col]:
        z = (i-mean)/std
        if z > threshold:
            outlier.append(i)
    print('les outliers pour la varible' + col + 'suivants : ', outlier)

les outliers pour la varibleIdsuivants :  []
les outliers pour la varibleMSSubClasssuivants :  [190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190, 190]
les outliers pour la varibleLotFrontagesuivants :  [141.0, 174.0, 174.0, 140.0, 150.0, 144.0, 149.0, 313.0, 168.0, 182.0, 160.0, 152.0, 313.0, 153.0, 150.0, 160.0, 195.0, 155.0, 149.0, 200.0, 140.0, 160.0]
les outliers pour la varibleLotAreasuivants :  [50271, 159000, 39104, 215245, 164660, 53107, 34650, 70761, 53227, 40094, 46589, 115149, 53504, 45600, 35760, 35133, 36500, 63887, 57200, 47280, 47007, 56600, 51974, 41600, 39290, 43500, 33983, 39384, 50102]
les outliers pour la varibleOverallQualsuivants :  []
les outliers pour la varibleOverallCondsuivants :  [9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9,

Nous avons ici, pour chacune de nos variables quantitatives les valeurs qui sont aberrantes. Pour traiter ces valeurs, nous avons plusieurs solutions :

    - supprimer les valeurs aberrantes
    - créer des classes pour atténuer l'impact des valeurs aberrantes
    - imputer les valeurs.
    
N'ayant pas beaucoup d'obeservations, nous avons choisis de créer des classes. Ainsi, nous évitons de perdre de l'information.

### Suppression de certaines variables : A FAIRE

Nous allons procéder à une première suppression des variables qui ne sont pas exploitables. Nous allons supprimer les variables ayant trop de valeurs manquantes, ou ayant une seule modalité ou une modalité qui revient presque à chaque fois comme par exemple avec la variable "Utilities" qui a une modlité présente 99.97% et une autre présente 0.03%.

In [14]:
df.drop(columns=["LotFrontage","Alley","FireplaceQu","PoolQC","Fence","MiscFeature"], inplace=True)

In [15]:
df2 = df

In [16]:
df2

Id  MSSubClass MSZoning  LotArea Street LotShape LandContour  \
0        1          60       RL     8450   Pave      Reg         Lvl   
1        2          20       RL     9600   Pave      Reg         Lvl   
2        3          60       RL    11250   Pave      IR1         Lvl   
3        4          70       RL     9550   Pave      IR1         Lvl   
4        5          60       RL    14260   Pave      IR1         Lvl   
...    ...         ...      ...      ...    ...      ...         ...   
1454  2915         160       RM     1936   Pave      Reg         Lvl   
1455  2916         160       RM     1894   Pave      Reg         Lvl   
1456  2917          20       RL    20000   Pave      Reg         Lvl   
1457  2918          85       RL    10441   Pave      Reg         Lvl   
1458  2919          60       RL     9627   Pave      Reg         Lvl   

     Utilities LotConfig LandSlope Neighborhood Condition1 Condition2  \
0       AllPub    Inside       Gtl      CollgCr       Norm       Norm   
1       AllPub       FR2       Gtl      Veenker      Feedr       Norm   
2       AllPub    Inside       Gtl      CollgCr       Norm       Norm   
3       AllPub    Corner       Gtl      Crawfor       Norm       Norm   
4       AllPub       FR2       Gtl      NoRidge       Norm       Norm   
...        ...       ...       ...          ...        ...        ...   
1454    AllPub    Inside       Gtl      MeadowV       Norm       Norm   
1455    AllPub    Inside       Gtl      MeadowV       Norm       Norm   
1456    AllPub    Inside       Gtl      Mitchel       Norm       Norm   
1457    AllPub    Inside       Gtl      Mitchel       Norm       Norm   
1458    AllPub    Inside       Mod      Mitchel       Norm       Norm   

     BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
0        1Fam     2Story            7            5       2003          2003   
1        1Fam     1Story            6            8       1976          1976   
2        1Fam     2Story            7            5       2001          2002   
3        1Fam     2Story            7            5       1915          1970   
4        1Fam     2Story            8            5       2000          2000   
...       ...        ...          ...          ...        ...           ...   
1454    Twnhs     2Story            4            7       1970          1970   
1455   TwnhsE     2Story            4            5       1970          1970   
1456     1Fam     1Story            5            7       1960          1996   
1457     1Fam     SFoyer            5            5       1992          1992   
1458     1Fam     2Story            7            5       1993          1994   

     RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea  \
0        Gable  CompShg     VinylSd     VinylSd    BrkFace       196.0   
1        Gable  CompShg     MetalSd     MetalSd       None         0.0   
2        Gable  CompShg     VinylSd     VinylSd    BrkFace       162.0   
3        Gable  CompShg     Wd Sdng     Wd Shng       None         0.0   
4        Gable  CompShg     VinylSd     VinylSd    BrkFace       350.0   
...        ...      ...         ...         ...        ...         ...   
1454     Gable  CompShg     CemntBd     CmentBd       None         0.0   
1455     Gable  CompShg     CemntBd     CmentBd       None         0.0   
1456     Gable  CompShg     VinylSd     VinylSd       None         0.0   
1457     Gable  CompShg     HdBoard     Wd Shng       None         0.0   
1458     Gable  CompShg     HdBoard     HdBoard    BrkFace        94.0   

     ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0           Gd        TA      PConc       Gd       TA           No   
1           TA        TA     CBlock       Gd       TA           Gd   
2           Gd        TA      PConc       Gd       TA           Mn   
3           TA        TA     BrkTil       TA       Gd           No   
4           Gd        TA      PConc       Gd       TA           Av   
...     

### Regroupement des modalitées

In [ ]:
# Regroupement de la variable âge
# On s'aide des Quartile et de la médiane pour les déterminer 
df_3.loc[df['age'] <= 28,'age'] = '17-28 ans'
df_3.loc[(df['age'] >= 29) & (df['age'] <= 37),'age'] = '29-37 ans'
df_3.loc[(df['age'] >= 38) & (df['age'] <= 48),'age'] = '38-48 ans'
df_3.loc[(df['age'] >= 49) & (df['age'] <= 60),'age'] = '49-60 ans'
df_3.loc[df['age'] >= 61,'age'] = '61 ans et +'

## 4. Modélisation 

### Échantillonnage 

In [ ]:
#On ne va pas travailler sur la totalité de nos observations, il est important de réaliser un échantillonnage 
#Train set, Test set
y = df['SalePrice'] #On défini la variable a expliquer
y=y.astype('int')
X = df.drop(columns = ["SalePrice"], inplace=True) # On défini nos features

### Dummies

In [ ]:
X = pd.get_dummies(X)
X

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, train_size=0.80, test_size=0.20, random_state=0, stratify=y) 
# On réalise l'échantillonage en splitant le datasets a 80% pour le train set et 20% pour le test set
# On initialise un random_state égal à 0 pour obtenir toujours le même tirage aléatoire
# On initialise le paramètre de stratification afin de conserver la répartition par strate